[New KB Article] Matching multiple tokens with ellipsis metavariables (#1140)
returntocorp/semgrep-docs: docs/kb/rules/ellipsis-metavariables.md

---
description: Ellipsis metavariables can help with matching multiple word tokens.
tags:
  - Rules
  - Semgrep Code
---

import MoreHelp from "/src/components/MoreHelp"

# Matching multiple tokens with ellipsis metavariables

> 

Reviewer: ```suggestion
Match multiple tokens, such as arguments of a function or elements of an array, through the use of two Semgrep rule constructs:

* Ellipsis metavariables
* Aliengrep
```

Just adding an intro and roughly defining a "token". Your subsequent definition of a token is not replaced.

Author: This isn't something I want to have upfront, in particular because this KB is about a less typical usage of ellipsis metavariables that isn't either of those things. But you have very quickly identified something that definitely isn't clear, and most of your comments basically center on this: what is a token?

Most of our docs only talk about tokens in the context of authentication (which is a special use/type of token), so we don't have a lot more on this, except in a few places. The [caveats section](https://semgrep.dev/docs/writing-rules/generic-pattern-matching/#caveats-and-limitations-of-generic-mode) that I link to is somewhat helpful:

> metavariable support is limited to capturing a single “word”, which is a token of the form [A-Za-z0-9_]+

Tokenization in language processing in general is a phase of analysis where you separate the language into its more basic elements. The details of the process, and what constitutes a token, depend on the language being analyzed and the purpose of the analysis. In the case of the generic language processor for Semgrep, it tokenizes `hello, world` as 3 tokens (as in the example), but only two of those tokens are "words"; the comma is a separator/punctuation. `ABC-DEF` is also broken down into 3 (two "words", ABC and DEF, and one separator, `-`), but not ABC_DEF - that's one token, because underscore is considered part of a word in that processor.

@emjin might have a good way to break this down more accessibly so we can present a brief definition that makes sense...

Reviewer: > This isn't something I want to have upfront, in particular because this KB is about a less typical usage of ellipsis metavariables that isn't either of those things.

Yeah, I guess the array / function example doesn't match the rest of the article. What would make a better example or use case in this case for generic "words"?

Author: I ended up reorganizing the content at the top and adding more clear definitions/examples of terms to try to resolve this. Hopefully it make sense. :)

> ## What are ellipsis metavariables?

Reviewer: ```suggestion
## What are ellipsis metavariables?
```

I'd suggest avoiding hypophora in technical writing, whether body text or headings. We use two types of headings:
* Task-based
* Conceptual

Alternatives
* Introduction to ellipsis metavariables
* Defining ellipsis metavariables

Author: Thanks, I adjusted it!

> Using ellipsis (`...`) to match a sequence of items is one of the most common constructs in Semgrep rules, and ellipses can even be used with metavariables ($VAR) to increase matching scope!

Reviewer: Are items the same as tokens?

Author: I think probably if we get tokens defined correctly above, yes. :)

Most commonly, ellipsis metavariables like `$...ARGS` are used for purposes like matching multiple arguments to a function, [while capturing the values for later re-use](/docs/writing-rules/pattern-syntax/#ellipsis-metavariables).

> However, they can also be used to match multiple word tokens. This is especially handy in [Generic pattern matching mode](/docs/writing-rules/generic-pattern-matching/), where a token is defined as a sequence of characters in the set `[A-z0-9_]`, which is not a fit for all languages.

Reviewer: ```suggestion
However, they can also be used to match multiple word tokens. This is especially handy in [Generic pattern matching mode](/docs/writing-rules/generic-pattern-matching/), where a token is defined as a sequence of characters in the set `[A-z0-9_]`, which is not a fit for all languages.
```

"Not a fit for all languages" - what are you trying to say here? I feel like there's a comparison being made between the `generic` language and other programming languages.

## Capturing multiple tokens with ellipsis metavariables

> In `generic` mode, `ABC_DEF` is one token, and a metavariable such as `$VAR` captures the entire sequence. However, `ABC-DEF` is two tokens, and a metavariable such as `$VAR` does not capture the entire sequence.

Author: Yeah, this is a version of the standard regex notation for the character class including all letters and digits, plus underscore. It's sometimes written [A-Za-z0-9] for clarity but that just makes it clearer that letters in both lower and uppercase are included.

It's actually not my definition, it comes from the generic mode caveat that I link to and which I quoted in the comment thread above. It doesn't include dash unless you add a dash after the underscore - a dash between two range endpoints in a character class can only be a range specifier, not a character on its own.

If the language you're working with allows other characters in tokens, using ellipsis metavariables can prevent problems with matching too little of the pattern.

To match `ABC-DEF` in `generic` mode, use an ellipsis metavariable, like `$...VAR`. Here is an example rule:

<iframe src="https://semgrep.dev/embed/editor?snippet=J6Ro" title="pattern-not rule for unverified transactions" width="100%" height="432px" frameBorder="0"></iframe>

If you remove the ellipsis in the `$...ID` variable, the second example no longer matches.

## Alternative: try the Aliengrep experiment

> Semgrep's generic matching mode is [somewhat limited](/docs/writing-rules/generic-pattern-matching/#caveats-and-limitations-of-generic-mode). To address some of the limitations, the team is experimenting with a new mode called [Aliengrep](/docs/writing-rules/experiments/aliengrep/). 

Reviewer: ```suggestion
Semgrep's generic matching mode contains [caveats and limitations](/docs/writing-rules/generic-pattern-matching/#caveats-and-limitations-of-generic-mode). To address some of the limitations, the team is experimenting with a new mode called [Aliengrep](/docs/writing-rules/experiments/aliengrep/). 
```
Just a suggestion to try to match the title of of the doc or section the user will go to when they click.
